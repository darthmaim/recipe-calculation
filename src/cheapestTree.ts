import { calculateTreeQuantity } from './calculateTreeQuantity'
import { calculateTreePrices } from './calculateTreePrices'
import { calculateTreeCraftFlags } from './calculateTreeCraftFlags'
import { RecipeTree, RecipeTreeWithCraftFlags, RecipeTreeWithPrices } from './types'
import {
  NestedRecipe,
  BasicCurrencyComponent,
  BasicItemComponent,
} from '@gw2efficiency/recipe-nesting'

export function cheapestTree(
  amount: number,
  tree: NestedRecipe,
  itemPrices: Record<string, number>,
  availableItems: Record<string, number> = {},
  forceBuyItems: Array<number> = [],
  valueOwnItems: false
): RecipeTreeWithCraftFlags {
  // calculateTreeQuantity already checks for craft flags, so we can set them here when valuing owned items
  if (valueOwnItems) {
    const treeWithQuantityWithoutAvailableItems = calculateTreeQuantity(
      amount,
      tree as RecipeTree,
      {}
    )
    const treeWithPriceWithoutAvailableItems = calculateTreePrices(
      treeWithQuantityWithoutAvailableItems,
      itemPrices
    )
    const cheaperToBuyItemIds = getCheaperToBuyItemIds(treeWithPriceWithoutAvailableItems)
    disableCraftForItemIds(tree, cheaperToBuyItemIds)
  }
  // Adjust the tree total and used quantities
  const treeWithQuantity = calculateTreeQuantity(amount, tree as RecipeTree, availableItems)

  // Set the initial craft flags based on the subtree prices
  const treeWithPrices = calculateTreePrices(treeWithQuantity, itemPrices)
  const treeWithCraftFlags = calculateTreeCraftFlags(treeWithPrices, forceBuyItems)

  // Force the root to be crafted
  treeWithCraftFlags.craft = true

  // After the "craft" flags are set, update the used materials
  // to only be used for things that actually get crafted
  const treeWithQuantityPostFlags = calculateTreeQuantity(
    amount,
    treeWithCraftFlags,
    availableItems
  )

  // Recalculate the correct tree price
  return calculateTreePrices(treeWithQuantityPostFlags, itemPrices)
}

function getCheaperToBuyItemIds(
  tree: RecipeTreeWithPrices,
  ids: Array<number> = []
): Array<number> {
  // If buying the item costs less than the possible profit made by selling its craft components (with TP tax), we should buy instead of crafting
  if (
    typeof tree.craftDecisionPrice === 'number' &&
    typeof tree.buyPrice === 'number' &&
    tree.buyPrice < tree.craftDecisionPrice * 0.85
  ) {
    if (!ids.includes(tree.id)) {
      ids.push(tree.id)
    }
  }

  if (tree.components && Array.isArray(tree.components)) {
    tree.components.forEach((component) => getCheaperToBuyItemIds(component, ids))
  }

  return ids
}

type NestedRecipeAndBasicComponentWithCraftFlag = (
  | NestedRecipe
  | BasicItemComponent
  | BasicCurrencyComponent
) & {
  craft?: boolean
}

function disableCraftForItemIds(
  tree: NestedRecipeAndBasicComponentWithCraftFlag,
  ids: Array<number>
) {
  if (ids.includes(tree.id)) {
    tree.craft = false
  }

  if ('components' in tree && Array.isArray(tree.components)) {
    tree.components.forEach((component) => {
      disableCraftForItemIds(component as NestedRecipeAndBasicComponentWithCraftFlag, ids)
    })
  }
}
