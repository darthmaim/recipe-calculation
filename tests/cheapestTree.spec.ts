// @ts-nocheck
import cheapestTree from '../src/cheapestTree'

describe('cheapestTree', () => {
  it('can calculate the cheapest tree correctly', () => {
    const recipeTree = {
      id: 1,
      quantity: 1,
      output: 1,
      components: [
        { id: 2, quantity: 1 },
        { id: 3, quantity: 5, components: [{ id: 4, quantity: 2 }] },
        { id: 5, quantity: 5, components: [{ id: 6, quantity: 2 }], recipe_id: 123 },
      ],
    }
    const prices = { 1: 10, 2: 42, 3: 10, 4: 10, 5: 1000, 6: 10 }

    const calculatedTree = cheapestTree(2, recipeTree, prices)
    expect(calculatedTree).toMatchSnapshot()
  })

  it('can calculate the cheapest tree correctly with available items', () => {
    const recipeTree = {
      id: 1,
      quantity: 1,
      output: 1,
      components: [
        { id: 7, quantity: 3, components: [{ id: 4, quantity: 1 }] },
        { id: 3, quantity: 5, components: [{ id: 4, quantity: 2 }] },
        { id: 5, quantity: 2, components: [{ id: 6, quantity: 100 }] },
      ],
    }
    const prices = { 1: 10, 3: 100, 4: 10, 5: 25, 6: 1, 7: 1 }
    const availableItems = { 1: 100, 3: 7, 4: 3, 5: 4 }

    const calculatedTree = cheapestTree(2, recipeTree, prices, availableItems, [])
    expect(calculatedTree).toMatchSnapshot()
  })

  it('can calculate the cheapest tree correctly with force buy items', () => {
    const recipeTree = {
      id: 1,
      quantity: 1,
      output: 1,
      components: [{ id: 3, quantity: 5, components: [{ id: 4, quantity: 2 }] }],
    }
    const prices = { 1: 10, 3: 100, 4: 10 }

    const calculatedTree = cheapestTree(2, recipeTree, prices, {}, [3])
    expect(calculatedTree).toMatchSnapshot()
  })
})
