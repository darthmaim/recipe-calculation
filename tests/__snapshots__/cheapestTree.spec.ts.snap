// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cheapestTree can calculate the cheapest tree correctly 1`] = `
Object {
  "buyPrice": 20,
  "buyPriceEach": 10,
  "components": Array [
    Object {
      "buyPrice": 84,
      "buyPriceEach": 42,
      "components": undefined,
      "craft": false,
      "decisionPrice": 84,
      "id": 2,
      "output": 1,
      "quantity": 1,
      "totalQuantity": 2,
      "usedQuantity": 2,
    },
    Object {
      "buyPrice": 100,
      "buyPriceEach": 10,
      "components": Array [
        Object {
          "buyPrice": 200,
          "buyPriceEach": 10,
          "components": undefined,
          "craft": false,
          "decisionPrice": 200,
          "id": 4,
          "output": 1,
          "quantity": 2,
          "totalQuantity": 20,
          "usedQuantity": 20,
        },
      ],
      "craft": false,
      "craftPrice": 200,
      "decisionPrice": 100,
      "id": 3,
      "output": 1,
      "quantity": 5,
      "totalQuantity": 10,
      "usedQuantity": 10,
    },
    Object {
      "buyPrice": 10000,
      "buyPriceEach": 1000,
      "components": Array [
        Object {
          "buyPrice": 200,
          "buyPriceEach": 10,
          "components": undefined,
          "craft": false,
          "decisionPrice": 200,
          "id": 6,
          "output": 1,
          "quantity": 2,
          "totalQuantity": 20,
          "usedQuantity": 20,
        },
      ],
      "craft": true,
      "craftPrice": 200,
      "decisionPrice": 200,
      "id": 5,
      "output": 1,
      "quantity": 5,
      "recipe_id": 123,
      "totalQuantity": 10,
      "usedQuantity": 10,
    },
  ],
  "craft": true,
  "craftPrice": 384,
  "decisionPrice": 384,
  "id": 1,
  "output": 1,
  "quantity": 1,
  "totalQuantity": 2,
  "usedQuantity": 2,
}
`;

exports[`cheapestTree can calculate the cheapest tree correctly with available items 1`] = `
Object {
  "buyPrice": 20,
  "buyPriceEach": 10,
  "components": Array [
    Object {
      "buyPrice": 6,
      "buyPriceEach": 1,
      "components": Array [
        Object {
          "buyPrice": 60,
          "buyPriceEach": 10,
          "components": undefined,
          "craft": false,
          "decisionPrice": 60,
          "id": 4,
          "output": 1,
          "quantity": 1,
          "totalQuantity": 6,
          "usedQuantity": 6,
        },
      ],
      "craft": false,
      "craftPrice": 60,
      "decisionPrice": 6,
      "id": 7,
      "output": 1,
      "quantity": 3,
      "totalQuantity": 6,
      "usedQuantity": 6,
    },
    Object {
      "buyPrice": 300,
      "buyPriceEach": 100,
      "components": Array [
        Object {
          "buyPrice": 30,
          "buyPriceEach": 10,
          "components": undefined,
          "craft": false,
          "decisionPrice": 30,
          "id": 4,
          "output": 1,
          "quantity": 2,
          "totalQuantity": 6,
          "usedQuantity": 3,
        },
      ],
      "craft": true,
      "craftPrice": 30,
      "decisionPrice": 30,
      "id": 3,
      "output": 1,
      "quantity": 5,
      "totalQuantity": 10,
      "usedQuantity": 3,
    },
    Object {
      "buyPrice": 0,
      "buyPriceEach": 25,
      "components": Array [
        Object {
          "buyPrice": 0,
          "buyPriceEach": 1,
          "components": undefined,
          "craft": false,
          "decisionPrice": 0,
          "id": 6,
          "output": 1,
          "quantity": 100,
          "totalQuantity": 0,
          "usedQuantity": 0,
        },
      ],
      "craft": false,
      "craftPrice": 0,
      "decisionPrice": 0,
      "id": 5,
      "output": 1,
      "quantity": 2,
      "totalQuantity": 4,
      "usedQuantity": 0,
    },
  ],
  "craft": true,
  "craftPrice": 36,
  "decisionPrice": 36,
  "id": 1,
  "output": 1,
  "quantity": 1,
  "totalQuantity": 2,
  "usedQuantity": 2,
}
`;

exports[`cheapestTree can calculate the cheapest tree correctly with force buy items 1`] = `
Object {
  "buyPrice": 20,
  "buyPriceEach": 10,
  "components": Array [
    Object {
      "buyPrice": 1000,
      "buyPriceEach": 100,
      "components": Array [
        Object {
          "buyPrice": 200,
          "buyPriceEach": 10,
          "components": undefined,
          "craft": false,
          "decisionPrice": 200,
          "id": 4,
          "output": 1,
          "quantity": 2,
          "totalQuantity": 20,
          "usedQuantity": 20,
        },
      ],
      "craft": false,
      "craftPrice": 200,
      "decisionPrice": 1000,
      "id": 3,
      "output": 1,
      "quantity": 5,
      "totalQuantity": 10,
      "usedQuantity": 10,
    },
  ],
  "craft": true,
  "craftPrice": 1000,
  "decisionPrice": 1000,
  "id": 1,
  "output": 1,
  "quantity": 1,
  "totalQuantity": 2,
  "usedQuantity": 2,
}
`;
